discovery/kubernetes/kubernetes.go

```diff
@@ -101,19 +122,70 @@ func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
 	if err != nil {
 		return err
 	}
-	if c.Role == "" {
-		return errors.Errorf("role missing (one of: pod, service, endpoints, node, ingress)")
-	}
 	err = c.HTTPClientConfig.Validate()
 	if err != nil {
 		return err
 	}
 	if c.APIServer.URL == nil && !reflect.DeepEqual(c.HTTPClientConfig, config_util.HTTPClientConfig{}) {
 		return errors.Errorf("to use custom HTTP client configuration please provide the 'api_server' URL explicitly")
 	}
+	switch c.Role {
```

Comment:
it could be shortened a bit using a map like this:

```go
var allowedSelectors = map[Role][]string{
   RolePod: []string{RolePod},
   RoleEndpoint: []string{RolePod, RoleEndpoint, RoleService},
   ...
}

...

for _, selector := range c.Selectors {
  if selector.Field != "" {
    break
  }
  var allowed bool
  for _, role := range allowedSelectors[c.Role] {
    if role == selector.Role {
      allowed = true
      break
    }
  }
  if !allowed {
    return errors.Errorf("%s role supports only %s selectors", c.Role, strings.Join(allowedSelectors[c.Role], ","))
  }
}
```